File: android/cyrebro_sdk/src/main/java/com/cyrebroplugin/LSLModule.kt
package com.cyrebroplugin

import android.util.Log
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.modules.core.DeviceEventManagerModule
import java.io.DataOutputStream
import java.net.InetAddress
import java.net.Socket
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

/**
 * LSL (Lab Streaming Layer) Module for React Native
 * Handles real-time EEG data streaming via LSL protocol
 */
class LSLModule(private val reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

    companion object {
        private const val TAG = "LSLModule"
        private const val LSL_DEFAULT_PORT = 16571
        private const val LSL_HEADER_SIZE = 16
        private const val LSL_CHUNK_SIZE = 8
    }

    override fun getName(): String = "LSLModule"

    private val executor: ScheduledExecutorService = Executors.newScheduledThreadPool(2)
    private val isStreaming = AtomicBoolean(false)
    private val activeOutlets = ConcurrentHashMap<String, LSLOutlet>()
    
    private var lslServerSocket: Socket? = null
    private var lslOutputStream: DataOutputStream? = null

    /**
     * LSL Outlet class to manage individual data streams
     */
    private inner class LSLOutlet(
        val name: String,
        val type: String,
        val channelCount: Int,
        val sampleRate: Double,
        val dataFormat: String
    ) {
        private var outletId: String = ""
        private var isActive = false
        private val buffer = mutableListOf<DoubleArray>()
        
        fun start(): Boolean {
            try {
                // Create LSL outlet info
                val outletInfo = createOutletInfo()
                outletId = "${name}_${System.currentTimeMillis()}"
                isActive = true
                
                Log.d(TAG, "LSL Outlet started: $name (ID: $outletId)")
                return true
            } catch (e: Exception) {
                Log.e(TAG, "Failed to start LSL outlet: $name", e)
                return false
            }
        }
        
        fun stop() {
            isActive = false
            buffer.clear()
            Log.d(TAG, "LSL Outlet stopped: $name")
        }
        
        fun pushSample(sample: DoubleArray) {
            if (isActive && sample.size == channelCount) {
                buffer.add(sample)
                
                // Send data in chunks to avoid blocking
                if (buffer.size >= LSL_CHUNK_SIZE) {
                    sendChunk()
                }
            }
        }
        
        private fun sendChunk() {
            if (buffer.isEmpty()) return
            
            try {
                val chunk = buffer.toTypedArray()
                buffer.clear()
                
                // Send chunk data via LSL protocol
                sendLSLChunk(chunk)
                
            } catch (e: Exception) {
                Log.e(TAG, "Failed to send LSL chunk for outlet: $name", e)
            }
        }
        
        private fun createOutletInfo(): String {
            return """
                {
                    "name": "$name",
                    "type": "$type",
                    "channel_count": $channelCount,
                    "sample_rate": $sampleRate,
                    "data_format": "$dataFormat",
                    "source_id": "$outletId"
                }
            """.trimIndent()
        }
    }

    /**
     * Initialize LSL server connection
     */
    @ReactMethod
    fun initializeLSL(serverAddress: String, port: Int, promise: Promise) {
        try {
            val address = if (serverAddress.isEmpty()) "localhost" else serverAddress
            val serverPort = if (port <= 0) LSL_DEFAULT_PORT else port
            
            Log.d(TAG, "Initializing LSL connection to $address:$serverPort")
            
            // Create socket connection to LSL server
            lslServerSocket = Socket(InetAddress.getByName(address), serverPort)
            lslOutputStream = DataOutputStream(lslServerSocket!!.getOutputStream())
            
            // Send initialization handshake
            sendLSLHandshake()
            
            promise.resolve(true)
            Log.d(TAG, "LSL connection established successfully")
            
        } catch (e: Exception) {
            Log.e(TAG, "Failed to initialize LSL connection", e)
            promise.reject("LSL_INIT_ERROR", "Failed to connect to LSL server: ${e.message}")
        }
    }

    /**
     * Create a new LSL outlet for streaming data
     */
    @ReactMethod
    fun createOutlet(
        name: String,
        type: String,
        channelCount: Int,
        sampleRate: Double,
        dataFormat: String,
        promise: Promise
    ) {
        try {
            if (name.isEmpty()) {
                promise.reject("INVALID_NAME", "Outlet name cannot be empty")
                return
            }
            
            if (activeOutlets.containsKey(name)) {
                promise.reject("OUTLET_EXISTS", "Outlet with name '$name' already exists")
                return
            }
            
            val outlet = LSLOutlet(name, type, channelCount, sampleRate, dataFormat)
            if (outlet.start()) {
                activeOutlets[name] = outlet
                promise.resolve(name)
                Log.d(TAG, "LSL outlet created: $name")
            } else {
                promise.reject("OUTLET_CREATE_ERROR", "Failed to create LSL outlet: $name")
            }
            
        } catch (e: Exception) {
            Log.e(TAG, "Error creating LSL outlet: $name", e)
            promise.reject("OUTLET_CREATE_ERROR", "Exception creating outlet: ${e.message}")
        }
    }

    /**
     * Push EEG sample data to LSL outlet
     */
    @ReactMethod
    fun pushSample(outletName: String, sample: ReadableArray, promise: Promise) {
        try {
            val outlet = activeOutlets[outletName]
            if (outlet == null) {
                promise.reject("OUTLET_NOT_FOUND", "LSL outlet '$outletName' not found")
                return
            }
            
            val sampleData = DoubleArray(sample.size())
            for (i in 0 until sample.size()) {
                sampleData[i] = sample.getDouble(i)
            }
            
            outlet.pushSample(sampleData)
            promise.resolve(true)
            
        } catch (e: Exception) {
            Log.e(TAG, "Error pushing sample to outlet: $outletName", e)
            promise.reject("PUSH_SAMPLE_ERROR", "Failed to push sample: ${e.message}")
        }
    }

    /**
     * Start streaming EEG data to LSL
     */
    @ReactMethod
    fun startStreaming(promise: Promise) {
        try {
            if (isStreaming.get()) {
                promise.resolve(true) // Already streaming
                return
            }
            
            isStreaming.set(true)
            
            // Start periodic data transmission
            executor.scheduleAtFixedRate({
                try {
                    flushAllOutlets()
                } catch (e: Exception) {
                    Log.e(TAG, "Error in streaming loop", e)
                }
            }, 0, 100, TimeUnit.MILLISECONDS) // 10Hz flush rate
            
            promise.resolve(true)
            Log.d(TAG, "LSL streaming started")
            
        } catch (e: Exception) {
            Log.e(TAG, "Failed to start LSL streaming", e)
            promise.reject("STREAMING_START_ERROR", "Failed to start streaming: ${e.message}")
        }
    }

    /**
     * Stop streaming EEG data to LSL
     */
    @ReactMethod
    fun stopStreaming(promise: Promise) {
        try {
            isStreaming.set(false)
            
            // Flush remaining data
            flushAllOutlets()
            
            promise.resolve(true)
            Log.d(TAG, "LSL streaming stopped")
            
        } catch (e: Exception) {
            Log.e(TAG, "Failed to stop LSL streaming", e)
            promise.reject("STREAMING_STOP_ERROR", "Failed to stop streaming: ${e.message}")
        }
    }

    /**
     * Remove an LSL outlet
     */
    @ReactMethod
    fun removeOutlet(outletName: String, promise: Promise) {
        try {
            val outlet = activeOutlets.remove(outletName)
            if (outlet != null) {
                outlet.stop()
                promise.resolve(true)
                Log.d(TAG, "LSL outlet removed: $outletName")
            } else {
                promise.reject("OUTLET_NOT_FOUND", "Outlet '$outletName' not found")
            }
            
        } catch (e: Exception) {
            Log.e(TAG, "Error removing LSL outlet: $outletName", e)
            promise.reject("OUTLET_REMOVE_ERROR", "Failed to remove outlet: ${e.message}")
        }
    }

    /**
     * Get list of active LSL outlets
     */
    @ReactMethod
    fun getActiveOutlets(promise: Promise) {
        try {
            val outlets = Arguments.createArray()
            activeOutlets.keys.forEach { name ->
                outlets.pushString(name)
            }
            promise.resolve(outlets)
            
        } catch (e: Exception) {
            Log.e(TAG, "Error getting active outlets", e)
            promise.reject("GET_OUTLETS_ERROR", "Failed to get outlets: ${e.message}")
        }
    }

    /**
     * Close LSL connection and cleanup
     */
    @ReactMethod
    fun closeLSL(promise: Promise) {
        try {
            // Stop streaming
            isStreaming.set(false)
            
            // Close all outlets
            activeOutlets.values.forEach { it.stop() }
            activeOutlets.clear()
            
            // Close socket connection
            lslOutputStream?.close()
            lslServerSocket?.close()
            
            lslOutputStream = null
            lslServerSocket = null
            
            promise.resolve(true)
            Log.d(TAG, "LSL connection closed")
            
        } catch (e: Exception) {
            Log.e(TAG, "Error closing LSL connection", e)
            promise.reject("LSL_CLOSE_ERROR", "Failed to close LSL: ${e.message}")
        }
    }

    /**
     * Send LSL handshake message
     */
    private fun sendLSLHandshake() {
        try {
            val handshake = """
                {
                    "type": "handshake",
                    "version": "1.0",
                    "protocol": "lsl",
                    "timestamp": ${System.currentTimeMillis()}
                }
            """.trimIndent()
            
            sendLSLMessage(handshake)
            
        } catch (e: Exception) {
            Log.e(TAG, "Failed to send LSL handshake", e)
        }
    }

    /**
     * Send LSL chunk data
     */
    private fun sendLSLChunk(chunk: Array<DoubleArray>) {
        try {
            val chunkSize = chunk.size
            val channelCount = if (chunk.isNotEmpty()) chunk[0].size else 0
            
            // Create chunk header
            val header = ByteBuffer.allocate(LSL_HEADER_SIZE).order(ByteOrder.LITTLE_ENDIAN)
            header.putInt(chunkSize)
            header.putInt(channelCount)
            header.putLong(System.currentTimeMillis())
            
            // Send header
            lslOutputStream?.write(header.array())
            
            // Send data
            chunk.forEach { sample ->
                sample.forEach { value ->
                    lslOutputStream?.writeDouble(value)
                }
            }
            
            lslOutputStream?.flush()
            
        } catch (e: Exception) {
            Log.e(TAG, "Failed to send LSL chunk", e)
        }
    }

    /**
     * Send LSL message
     */
    private fun sendLSLMessage(message: String) {
        try {
            val messageBytes = message.toByteArray(Charsets.UTF_8)
            val length = messageBytes.size
            
            // Send message length
            lslOutputStream?.writeInt(length)
            
            // Send message content
            lslOutputStream?.write(messageBytes)
            lslOutputStream?.flush()
            
        } catch (e: Exception) {
            Log.e(TAG, "Failed to send LSL message", e)
        }
    }

    /**
     * Flush all active outlets
     */
    private fun flushAllOutlets() {
        activeOutlets.values.forEach { outlet ->
            try {
                // Force flush any remaining data
                outlet.pushSample(DoubleArray(outlet.channelCount) { 0.0 })
            } catch (e: Exception) {
                Log.e(TAG, "Error flushing outlet: ${outlet.name}", e)
            }
        }
    }

    override fun onCatalystInstanceDestroy() {
        super.onCatalystInstanceDestroy()
        
        try {
            // Cleanup resources
            isStreaming.set(false)
            activeOutlets.values.forEach { it.stop() }
            activeOutlets.clear()
            
            lslOutputStream?.close()
            lslServerSocket?.close()
            
            executor.shutdown()
            
        } catch (e: Exception) {
            Log.e(TAG, "Error during cleanup", e)
        }
    }
} 

File: ios/LSLModule.swift
import Foundation
import React
import Network

/**
 * LSL (Lab Streaming Layer) Module for React Native iOS
 * Handles real-time EEG data streaming via LSL protocol
 */
@objc(LSLModule)
class LSLModule: NSObject {
    
    private static let TAG = "LSLModule"
    private static let LSL_DEFAULT_PORT: UInt16 = 16571
    private static let LSL_HEADER_SIZE = 16
    private static let LSL_CHUNK_SIZE = 8
    
    private var executor: DispatchQueue?
    private var isStreaming = false
    private var activeOutlets: [String: LSLOutlet] = [:]
    
    private var lslConnection: NWConnection?
    private var lslOutputStream: OutputStream?
    
    /**
     * LSL Outlet class to manage individual data streams
     */
    private class LSLOutlet {
        let name: String
        let type: String
        let channelCount: Int
        let sampleRate: Double
        let dataFormat: String
        
        private var outletId: String = ""
        private var isActive = false
        private var buffer: [[Double]] = []
        
        init(name: String, type: String, channelCount: Int, sampleRate: Double, dataFormat: String) {
            self.name = name
            self.type = type
            self.channelCount = channelCount
            self.sampleRate = sampleRate
            self.dataFormat = dataFormat
        }
        
        func start() -> Bool {
            do {
                // Create LSL outlet info
                let outletInfo = createOutletInfo()
                outletId = "\(name)_\(Int(Date().timeIntervalSince1970 * 1000))"
                isActive = true
                
                print("\(LSLModule.TAG): LSL Outlet started: \(name) (ID: \(outletId))")
                return true
            } catch {
                print("\(LSLModule.TAG): Failed to start LSL outlet: \(name), error: \(error)")
                return false
            }
        }
        
        func stop() {
            isActive = false
            buffer.removeAll()
            print("\(LSLModule.TAG): LSL Outlet stopped: \(name)")
        }
        
        func pushSample(sample: [Double]) {
            if isActive && sample.count == channelCount {
                buffer.append(sample)
                
                // Send data in chunks to avoid blocking
                if buffer.count >= LSL_CHUNK_SIZE {
                    sendChunk()
                }
            }
        }
        
        private func sendChunk() {
            if buffer.isEmpty { return }
            
            do {
                let chunk = buffer
                buffer.removeAll()
                
                // Send chunk data via LSL protocol
                sendLSLChunk(chunk)
                
            } catch {
                print("\(LSLModule.TAG): Failed to send LSL chunk for outlet: \(name), error: \(error)")
            }
        }
        
        private func createOutletInfo() -> String {
            return """
            {
                "name": "\(name)",
                "type": "\(type)",
                "channel_count": \(channelCount),
                "sample_rate": \(sampleRate),
                "data_format": "\(dataFormat)",
                "source_id": "\(outletId)"
            }
            """
        }
        
        private func sendLSLChunk(_ chunk: [[Double]]) {
            // Implementation for sending LSL chunk data
            // This would involve creating the proper LSL protocol headers and data
            print("\(LSLModule.TAG): Sending LSL chunk for outlet: \(name), samples: \(chunk.count)")
        }
    }
    
    /**
     * Initialize LSL server connection
     */
    @objc(initializeLSL:port:withResolver:withRejecter:)
    func initializeLSL(serverAddress: String, port: Int, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        let address = serverAddress.isEmpty ? "localhost" : serverAddress
        let serverPort = port <= 0 ? Int(LSLModule.LSL_DEFAULT_PORT) : port
        
        print("\(LSLModule.TAG): Initializing LSL connection to \(address):\(serverPort)")
        
        // Create network connection to LSL server
        let endpoint = NWEndpoint.hostPort(host: NWEndpoint.Host(address), port: NWEndpoint.Port(integerLiteral: UInt16(serverPort)))
        lslConnection = NWConnection(to: endpoint, using: .tcp)
        
        lslConnection?.stateUpdateHandler = { [weak self] state in
            switch state {
            case .ready:
                print("\(LSLModule.TAG): LSL connection established successfully")
                resolve(true)
            case .failed(let error):
                print("\(LSLModule.TAG): LSL connection failed: \(error)")
                reject("LSL_INIT_ERROR", "Failed to connect to LSL server: \(error)", error)
            case .cancelled:
                print("\(LSLModule.TAG): LSL connection cancelled")
                reject("LSL_INIT_ERROR", "LSL connection was cancelled", nil)
            default:
                break
            }
        }
        
        lslConnection?.start(queue: .global())
        
        // Send initialization handshake
        sendLSLHandshake()
    }
    
    /**
     * Create a new LSL outlet for streaming data
     */
    @objc(createOutlet:type:channelCount:sampleRate:dataFormat:withResolver:withRejecter:)
    func createOutlet(name: String, type: String, channelCount: Int, sampleRate: Double, dataFormat: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        if name.isEmpty {
            reject("INVALID_NAME", "Outlet name cannot be empty", nil)
            return
        }
        
        if activeOutlets[name] != nil {
            reject("OUTLET_EXISTS", "Outlet with name '\(name)' already exists", nil)
            return
        }
        
        let outlet = LSLOutlet(name: name, type: type, channelCount: channelCount, sampleRate: sampleRate, dataFormat: dataFormat)
        if outlet.start() {
            activeOutlets[name] = outlet
            resolve(name)
            print("\(LSLModule.TAG): LSL outlet created: \(name)")
        } else {
            reject("OUTLET_CREATE_ERROR", "Failed to create LSL outlet: \(name)", nil)
        }
    }
    
    /**
     * Push EEG sample data to LSL outlet
     */
    @objc(pushSample:sample:withResolver:withRejecter:)
    func pushSample(outletName: String, sample: [Double], resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let outlet = activeOutlets[outletName] else {
            reject("OUTLET_NOT_FOUND", "LSL outlet '\(outletName)' not found", nil)
            return
        }
        
        outlet.pushSample(sample: sample)
        resolve(true)
    }
    
    /**
     * Start streaming EEG data to LSL
     */
    @objc(startStreaming:withRejecter:)
    func startStreaming(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        if isStreaming {
            resolve(true) // Already streaming
            return
        }
        
        isStreaming = true
        
        // Start periodic data transmission
        executor = DispatchQueue(label: "LSLStreamingQueue")
        executor?.async { [weak self] in
            while self?.isStreaming == true {
                self?.flushAllOutlets()
                Thread.sleep(forTimeInterval: 0.1) // 10Hz flush rate
            }
        }
        
        resolve(true)
        print("\(LSLModule.TAG): LSL streaming started")
    }
    
    /**
     * Stop streaming EEG data to LSL
     */
    @objc(stopStreaming:withRejecter:)
    func stopStreaming(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        isStreaming = false
        
        // Flush remaining data
        flushAllOutlets()
        
        resolve(true)
        print("\(LSLModule.TAG): LSL streaming stopped")
    }
    
    /**
     * Remove an LSL outlet
     */
    @objc(removeOutlet:withResolver:withRejecter:)
    func removeOutlet(outletName: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let outlet = activeOutlets.removeValue(forKey: outletName) else {
            reject("OUTLET_NOT_FOUND", "Outlet '\(outletName)' not found", nil)
            return
        }
        
        outlet.stop()
        resolve(true)
        print("\(LSLModule.TAG): LSL outlet removed: \(outletName)")
    }
    
    /**
     * Get list of active LSL outlets
     */
    @objc(getActiveOutlets:withRejecter:)
    func getActiveOutlets(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        let outlets = Array(activeOutlets.keys)
        resolve(outlets)
    }
    
    /**
     * Close LSL connection and cleanup
     */
    @objc(closeLSL:withRejecter:)
    func closeLSL(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        // Stop streaming
        isStreaming = false
        
        // Close all outlets
        activeOutlets.values.forEach { $0.stop() }
        activeOutlets.removeAll()
        
        // Close network connection
        lslConnection?.cancel()
        lslConnection = nil
        
        resolve(true)
        print("\(LSLModule.TAG): LSL connection closed")
    }
    
    /**
     * Send LSL handshake message
     */
    private func sendLSLHandshake() {
        let handshake = """
        {
            "type": "handshake",
            "version": "1.0",
            "protocol": "lsl",
            "timestamp": \(Int(Date().timeIntervalSince1970 * 1000))
        }
        """
        
        sendLSLMessage(handshake)
    }
    
    /**
     * Send LSL message
     */
    private func sendLSLMessage(_ message: String) {
        guard let data = message.data(using: .utf8) else { return }
        
        lslConnection?.send(content: data, completion: .contentProcessed { error in
            if let error = error {
                print("\(LSLModule.TAG): Failed to send LSL message: \(error)")
            }
        })
    }
    
    /**
     * Flush all active outlets
     */
    private func flushAllOutlets() {
        activeOutlets.values.forEach { outlet in
            // Force flush any remaining data
            outlet.pushSample(sample: Array(repeating: 0.0, count: outlet.channelCount))
        }
    }
}

@objc(LSLModule)
extension LSLModule: RCTBridgeModule {
    static func moduleName() -> String! {
        return "LSLModule"
    }
    
    static func requiresMainQueueSetup() -> Bool {
        return false
    }
} 